src/lidar/read_rp_lidar.cpp

```diff
@@ -12,60 +12,84 @@
 
 using namespace navtypes;
 
-std::atomic<bool> lidar_initialized(false);
 constexpr uint32_t RPLIDAR_BAUDRATE = 115200;
+const double MM_TO_M = 1000;
+
+std::atomic<bool> lidar_initialized(false);
+std::atomic<datatime_t> lidar_time(datatime_t{});
 RPLidar rp_lidar("/dev/ttyUSB0", RPLIDAR_BAUDRATE);
 std::thread lidar_thread;
-points_t last_points = {};
+DataPoint<points_t> last_points = {};
 std::mutex points_lock;
 
 namespace lidar {
 
-bool initializeLidar() {
-    if (!lidar_initialized) {
-        if (!rp_lidar.checkHealth()) {
-            perror("failed to connect to rp lidar");
-        }
-        rp_lidar.setMaxDistance(16.0);
-        lidar_thread = std::thread(&readLidarLoop);
-    }
-    return lidar_initialized;
-}
-
+/**
+ * @brief RP Lidar continuously scans environment, updates current timeframe with lidar data
+ */
 void readLidarLoop() {
     using namespace std::chrono;
     std::vector<Polar2D> polarPts;    
     while (true) {
         auto scan = rp_lidar.poll();
         if (auto scan = rp_lidar.poll()) {
-            // Converts data into Polar Coord
-            std::vector<Polar2D> currFrames;
+            std::vector<point_t> pts;
             double dtheta = (scan.value().angle_max-scan.value().angle_min)/(scan.value().ranges.size()-1);
             for (unsigned long i = 0; i < scan.value().ranges.size(); i++) {
                 double rad = dtheta*i;
-                double dist = scan.value().ranges[i] * 1000;
+                double dist = scan.value().ranges[i] * MM_TO_M;
 
                 Polar2D frame{dist, rad};
-                currFrames.push_back(frame);
+                pts.push_back(lidar::polarToCartesian2(frame));
             }
         
-            // Converts to regular points
-            std::vector<point_t> pts(currFrames.size());
-            point_t origin({0, 0, 1});
-            for (unsigned long i = 0; i < currFrames.size(); i++) {
-                pts[i] = lidar::polarToCartesian2(currFrames[i]);
-            }
             points_lock.lock();
-            last_points = pts;
+            datatime_t time = lidar_time;
+            last_points = {time, pts};
             points_lock.unlock();
-            lidar_initialized = true;
         } else {
             perror("failed to get frame");
+            points_lock.lock();
+            datatime_t time = lidar_time;
+            last_points = {};
+            points_lock.unlock();
```

Comment:
If the lidar fails to get a frame it might make sense to leave `last_points` as its current value. Clients will see that the timestamp hasn't changed so they'll know that the data is stale.

If we set an empty data point then the previous data is no longer available, even if it's still relatively new.